src/shared_fs.py

```diff
@@ -294,4 +320,8 @@ def move_to_quarantine(runner_name: str) -> None:
             f"Failed to archive shared filesystem for runner {runner_name}"
         ) from exc
 
-    delete(runner_name)
+    try:
+        delete(runner_name)
+    except DeleteSharedFilesystemError as exc:
+        logger.error("Error deleting shared filesystem, %s", exc)
+        raise
```

Comment:
I think this will lead to duplicate logs (e.g. the caller in runner_metrics will also log the exception (with trace)). I have seen this pattern (catch exc, log and reraise) in several places in this PR and I wonder if we should not generally let the caller log such exceptions. The callee should throw the exception and let the caller do what they want with it. Thoughts @jdkandersson @gregory-schiano ?

Reply:
Right, this was done to please the linter (re-raise without doing anything to exc). I think we may want to have a linter that lets `Raises` section without explicitly needing to raise the exception. (cc @yhaliaw , I think you mentioned there weren't quite a linter to make sure all exceptions are covered, something along the context here that you might be interested in)

Reply:
I am not sure if we should add code complexity (try catch raise) just to please the linter. I think the linting rules come from https://github.com/jdkandersson/flake8-docstrings-complete#fix-dco051, which is related to https://jdkandersson.com/2023/01/07/writing-great-docstrings-in-python/, which says not to include exceptions in the docstring if they come from a function within a function:

>  I tend not to include exceptions raised by functions called within the function since that can explode in complexity. The function should gracefully handle those exceptions anyway. 

So maybe remove the exceptions from the docstring and these constructs at least for functions marked private (so not part of the public module interface)?

Reply:
Well I tend to agree with Christopher, we should have constructs that try/except/raise just to please the linter, worst case we ignore the linter rule.
In my opinion the only condition a try/except/raise block should occur is when we want to wrap the exception in a custom error (which should be systematically done when the exception is a system or a library exception).
I agree that documenting all the underlying exception is complex and could lead to LOTs of exception, hence wrapping the relevant one in custom errors as parent custom error is an error to be excepted potentially anywhere in our application.

Reply:
Yeh just re-raising is generally not great. What can be reasonable is to catch and raise a different exception because the module has additional context to add to the exception. For example, at the low level, you might get an exception raised around that a file doesn't exists. Then the module might handle that exception and put it into context, such as that there was an error creating the configuration file. Then it is valuable to re-map to provide that extra context.

Log messages again can be useful if there is unique context that can only be added by a given function

I generally don't like letting exceptions bubble up because how is the caller supposed to know that they should handle it?

Reply:
How about re-raising it and removing the logs but silencing the linter to avoid duplicate logs?
I do agree that another exception should be defined for each module and let the caller handle those exceptions, perhaps in the refactor PR we should make sure that is the case. I don't mind defining the exceptions here either but I'm just afraid that it might introduce a lot of merge conflicts with Openstack PR.

Reply:
Perhaps initially just do it in the code being changed and then a follow up PR can do the rest?

Reply:
Updated - all direct re-raises are now pylint-ignored with comment.
```
# 2024/04/02 - We should define a new error, wrap it and re-raise it.
```
to be fixed in a followup PR.
Ticket for followup has been created with the title "GitHub self-hosted runner: code style issues fix for direct re-raises".